### PR TITLE
Adding field authorization

### PR DESCRIPTION
**Pull request by CNAChino**
# What Changed
Add field level authorization capability.  This uses a query visitor to check if access allowed.

# Why
Applications requires field level authorization

- [x] Add tests
- [ ] Add docs

### CODE REVIEW

File: src/main/java/com/intuit/graphql/orchestrator/authorization/FieldAuthorization.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+public interface FieldAuthorization<AuthDataT> {
+
+    boolean isAccessAllowed(FieldAuthorizationRequest<AuthDataT> fieldAuthorizationRequest);
+
+    boolean requiresAccessControl(FieldPosition fieldPosition);
+
+    CompletableFuture<AuthDataT> getFutureAuthData();
+
+    GraphqlErrorException getDeniedGraphQLErrorException();
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/authorization/FieldAuthorization.java:15]: Any reason for this method to be implemented by the class overwriting it.

File: src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java [hunk 2]
```diff
@@ -46,8 +44,13 @@
...
 import org.apache.commons.lang3.StringUtils;
 import org.dataloader.BatchLoader;
+import org.apache.commons.collections4.CollectionUtils;
+import com.intuit.graphql.orchestrator.authorization.DeclinedField;
+import com.intuit.graphql.orchestrator.authorization.FieldAuthorization;
+import com.intuit.graphql.orchestrator.authorization.QueryRedactor;
+import graphql.analysis.QueryTransformer;
 
-public class GraphQLServiceBatchLoader implements BatchLoader<DataFetchingEnvironment, DataFetcherResult<Object>> {
+public class GraphQLServiceBatchLoader<AuthDataT> implements BatchLoader<DataFetchingEnvironment, DataFetcherResult<Object>> {
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java:53]: I dont think this class needs to be generic. Does it give a compilation error?

File: src/main/java/com/intuit/graphql/orchestrator/common/FieldPosition.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Objects;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+@Getter
+public class FieldPosition {
+
+    @EqualsAndHashCode.Include
+    private final String parentTypename;
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/common/FieldPosition.java:13]: You can use graphql-javas FieldCoordinate class here for parent and fieldname

File: src/main/java/com/intuit/graphql/orchestrator/authorization/FieldAuthorization.java (hunk 1 above)
Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/authorization/FieldAuthorization.java:15]: The objective is to 
- allow the implementing class create custom GraphQLErrorExceptions is required by the application e.g. adding error codes in `error.extenstions`
- ensure that the error is thrown.  Was thinking if the actual throwing of exception should be performed by the implement class or the BatchLoader/Query Redactor.

File: src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java (hunk 2 above)
Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java:52]: Yes it gives compile error without the Generic Type.

File: src/main/java/com/intuit/graphql/orchestrator/common/FieldPosition.java (hunk 3 above)
Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/common/FieldPosition.java:12]: yup this is done

Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/common/FieldPosition.java]: this file is removed

File: src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java [hunk 4]
```diff
@@ -73,7 +73,24 @@ private GraphQLServiceBatchLoader(Builder builder) {
   @SuppressWarnings("rawtypes")
   @Override
   public CompletionStage<List<DataFetcherResult<Object>>> load(final List<DataFetchingEnvironment> keys) {
-    GraphQLContext context = getContext(keys);
+    GraphQLContext graphQLContext = getContext(keys);
+    FieldAuthorization<AuthDataT> fieldAuthorization = getFieldAuthorizationFromOrDefault(graphQLContext);
+    if (Objects.nonNull(fieldAuthorization)) {
+      CompletableFuture<AuthDataT> futureAuthData = fieldAuthorization.getFutureAuthData();
+      return futureAuthData.thenCompose(authData -> load(keys, authData, graphQLContext, fieldAuthorization));
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/batch/GraphQLServiceBatchLoader.java:80]: need to further optimize this as this always tries to complete the future even if the the data fetching environment selection set does not require authorization.

**Pull request closed without merging**